tailnet/multiagent.go

```diff
@@ -0,0 +1,179 @@
+package tailnet
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/uuid"
+
+	"cdr.dev/slog"
+)
+
+type MultiAgentConn interface {
+	UpdateSelf(node *Node) error
+	SubscribeAgent(agentID uuid.UUID) (func(), error)
+	UnsubscribeAgent(agentID uuid.UUID)
+	NextUpdate(ctx context.Context) []*Node
+	AgentIsLegacy(agentID uuid.UUID) bool
+	Close() error
+	IsClosed() bool
+}
+
+type MultiAgent struct {
+	mu     sync.RWMutex
+	closed chan struct{}
+
+	ID     uuid.UUID
+	Logger slog.Logger
+
+	AgentIsLegacyFunc func(agentID uuid.UUID) bool
+	OnSubscribe       func(enq Enqueueable, agent uuid.UUID) (close func(), err error)
+	OnNodeUpdate      func(id uuid.UUID, agents []uuid.UUID, node *Node) error
+
+	updates          chan []*Node
+	subscribedAgents map[uuid.UUID]func()
+}
+
+func (m *MultiAgent) Init() *MultiAgent {
+	m.closed = make(chan struct{})
+	m.updates = make(chan []*Node, 128)
+	m.subscribedAgents = map[uuid.UUID]func(){}
+	return m
+}
+
+func (m *MultiAgent) AgentIsLegacy(agentID uuid.UUID) bool {
+	return m.AgentIsLegacyFunc(agentID)
+}
+
+func (m *MultiAgent) UpdateSelf(node *Node) error {
+	m.mu.Lock()
+	agents := make([]uuid.UUID, 0, len(m.subscribedAgents))
+	for agent := range m.subscribedAgents {
+		agents = append(agents, agent)
+	}
+	m.mu.Unlock()
+
+	return m.OnNodeUpdate(m.ID, agents, node)
+}
+
+func (m *MultiAgent) SubscribeAgent(agentID uuid.UUID) (func(), error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if closer, ok := m.subscribedAgents[agentID]; ok {
+		return closer, nil
+	}
+
+	closer, err := m.OnSubscribe(m.enqueuer(agentID), agentID)
+	if err != nil {
+		return nil, err
+	}
+	m.subscribedAgents[agentID] = closer
+	return closer, nil
+}
+
+func (m *MultiAgent) UnsubscribeAgent(agentID uuid.UUID) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if closer, ok := m.subscribedAgents[agentID]; ok {
+		closer()
+	}
+	delete(m.subscribedAgents, agentID)
+}
+
+func (m *MultiAgent) NextUpdate(ctx context.Context) []*Node {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case nodes := <-m.updates:
+			return nodes
+		}
+	}
+}
+
+func (m *MultiAgent) enqueuer(agentID uuid.UUID) Enqueueable {
+	return &multiAgentEnqueuer{
+		agentID: agentID,
+		m:       m,
+	}
+}
+
+type multiAgentEnqueuer struct {
+	m *MultiAgent
+
+	agentID    uuid.UUID
```

Comment:
I guess I was mostly trying to avoid footguns when closing, and also allow for separate stat tracking. It seemed confusing/surprising to me that closing a single `Enqueuable` could have effects beyond that instance, possibly closing other `Enqueuable`s, especially when we're trying to treat them as equals to regular connections.

Reply:
But, what does it mean to "close" one of these?

Closing a queue should mean that you can't enqueue anything else on it and that the other side should get notified that it's closed.  That needs to be done at the level of the `MultiAgent` as a whole, not just in the context of receiving updates for a single agent.